src/abstractdataframe/abstractdataframe.jl

```diff
@@ -763,7 +763,7 @@ Base.hcat(df1::AbstractDataFrame, df2::AbstractDataFrame, dfn::AbstractDataFrame
 
 @generated function promote_col_type(cols::AbstractVector...)
     T = mapreduce(x -> Missings.T(eltype(x)), promote_type, cols)
-    if T <: CategoricalValue
+    if T <: CategoricalValue && T !== Union{}
```

Comment:
You don't have to change this, I just want to note that this can be written as `Union{} <: T <: CategoricalValue`. Kind of fun. :slightly_smiling_face:

Reply:
Hmm, on Julia 0.6.2:
```
julia> using CategoricalArrays

julia> Union{} <: Union{} <: CategoricalString
true

julia> Union{} <: CategoricalString{UInt32} <: CategoricalValue
false
```

Note the last comparison. It's confusing and unfortunate, but there could be categorical values types that are not `<: CategoricalValue` (in fact, `CategoricalString` is more frequent than `CategoricalValue`).
So the proper check is `CategoricalArrays.iscatvalue(T)`.